eth/db/header.py

```diff
@@ -240,6 +287,47 @@ def _persist_header_chain(
 
         return tuple(), tuple()
 
+    @classmethod
+    def _handle_gap_change(cls,
+                           db: DatabaseAPI,
+                           gap_change: GapChange,
+                           header: BlockHeaderAPI,
+                           genesis_parent_hash: Hash32) -> None:
+
+        if gap_change not in GAP_WRITES:
+            return
+
+        if gap_change is GapChange.GapFill:
+            expected_child = cls._get_canonical_head(db)
+            if header.hash != expected_child.parent_hash:
+                # We are trying to close a gap with an uncle. Reject!
+                raise ValidationError(f"{header} is not the parent of {expected_child}")
+
+        # We implicitly assert that persisted headers are canonical here.
+        # This assertion is made when persisting headers that are known to be part of a gap
+        # in the canonical chain.
+        # What we don't know is if this header will eventually lead up to the upper end of the
+        # gap (the checkpoint header) or if this is an alternative history. But we do ensure the
+        # integrity eventually. For one, we check the final header that would close the gap and if
+        # it does not match our expected child (the checkpoint header), we will reject it.
+        # Additionally, if we have persisted a chain of alternative history under the wrong
+        # assumption that it would be the canonical chain, but then at a later point it turns out it
+        # isn't and we overwrite it with the correct canonical chain, we make sure to
+        # recanonicalize all affected headers.
+        # IMPORTANT: Not only do we have to take this into account when we fill a gap, but also
+        # every time we write into a gap.
+        # Suppose we have a gap of 10 headers, then we fill 1 -5 with uncles from chain B.
+        # Now suppose we find the original chain A and attempt to write headers 1 - 10. At that
+        # point, we are under the assumption our current gap spans from just 6 - 10 because we have
+        # previously written headers 1 - 5 from chain B *believing* they are canonical which they
+        # turn out not to be. That means by the time we write headers 1 -5 again (but from chain A)
+        # we do not recognize it as writing to a known gap. Therefore by the time we write header 6,
+        # when we finally realize that we are attempting to fill in a gap, we have to backtrack
+        # the ancestors to add them to the canonical chain.
```

Comment:
This is quite a wall of text but it helped me understand why we we always want the call to `_find_new_ancestors` here.

Reply:
Hm, well, I think we can drop the last paragraph (or at least move it to the test), since it's basically just a test description, and I think it's the test you already wrote. (Could reference `test_gap_filling_handles_uncles_correctly` in a comment here if you want)